### PR TITLE
killing the process and child process on timeout

### PR DESCRIPTION
**Pull request by varun06**
With current version we only kill the parent process
on timeout. That makes all the child process orphaned.
This generally make the build fail, I have been facing
this issue a lot with keryx. So now I create a process
group and kill the group on timout. I modified the test
also.

### CODE REVIEW

File: execute.go [hunk 1]
```diff
@@ -84,23 +85,28 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
+		if err == nil {
+			if err := syscall.Kill(-processGroupID, syscall.SIGKILL); err != nil {
+				return 0, fmt.Errorf("Failed to kill process: %v", cmd.Process.Pid, err)
+			}
+		} else {
+			return 0, fmt.Errorf("Failed to get process group id: %v", cmd.Process.Pid, err)
 		}
 		<-done
-		exitCode = -23
+		//exitCode = -23
```
Comment (kklipsch) [inline execute.go]: meant to be commented out? If so delete.

**Comment by kklipsch** [comment]
You should write a test that creates a grandchild process and kills it to verify that this works.  This test has been needed since the first time we put time out logic in, so if you are coming back to this code now you might as well write it.

File: execute.go (hunk 1 above)
Comment (varun06) [inline execute.go]: I was going to ask you, why you were returning -23?

**Comment by varun06** [comment]
I modified that test to create a child process, I have hopes that it works coz I spent some time this weekend to learn a bit about Go process handling. It could be because of a bug in std lib where exec leaks processes but don't know. If it doesn't work, I will try a solution that goes to *os.Process to gather and kill child/grandchilds.

I will modify the test name.

File: execute.go (hunk 1 above)
Comment (kklipsch) [inline execute.go]: its not a used exit code.

File: execute.go [hunk 2]
```diff
@@ -84,23 +85,28 @@ func executeWithOutputChan(outputChan chan string, env []string, workingDir stri
...
+		done <- exitErr
 	}()
 
 	select {
 	case <-time.After(timeout):
-		if err := cmd.Process.Kill(); err != nil {
-			return 0, fmt.Errorf("Failed to kill process: %v", err)
+		processGroupID, err := syscall.Getpgid(cmd.Process.Pid)
+		if err == nil {
+			if err := syscall.Kill(-processGroupID, syscall.SIGKILL); err != nil {
```
Comment (bhand-mm) [inline execute.go]: Why `-processGroupID `?

File: execute.go [hunk 3]
```diff
@@ -112,17 +118,23 @@ func killProcessOnTimeout(cmd *exec.Cmd, timeout time.Duration) (int, error) {
 	return exitCode, nil
 }
 
-// gets the exit code from error
+func processExitCode(err error) int {
+	var exitCode int
+	if err != nil {
+		var exitErr error
+		if exitCode, exitErr = getExitCode(err); exitErr != nil {
+			exitCode = 127
```
Comment (bhand-mm) [inline execute.go]: Why 127?

File: execute_test.go [hunk 4]
```diff
@@ -88,7 +88,10 @@ func TestRunEchoWithChan(t *testing.T) {
 func TestGetExitCode(t *testing.T) {
 	timeoutTime := time.Duration(1) * time.Second
 
-	cmd := exec.Command("grep", "go")
+	//This command creates a parent and child process
+	//so that we can test killing both parent and child
+	//processes
+	cmd := exec.Command("/bin/sh", "-c", "watch date > date.txt")
```
Comment (bhand-mm) [inline execute_test.go]: What about a command that creates a process that creates a process?  What about a command that creates two processes that each create two processes?  I understand the permutations are infinite but it does seem like we should go at least a few levels deep to cover things like bash -> make -> bash -> postgres -> forked postgres's etc

File: execute.go (hunk 2 above)
Comment (varun06) [inline execute.go]: that's to target the group id instead of PID.

File: execute.go (hunk 3 above)
Comment (varun06) [inline execute.go]: This need more context instead of returning 127. I will make some changes after getting this kill process thing.

**Comment by varun06** [comment]
Added one more test to create process tree with grand child too. tests are passing and I can see processes getting killed on my machine.

File: execute.go (hunk 2 above)
Comment (kklipsch) [inline execute.go]: the calling convention is to kill the "negative" processGroupId?

**Comment by varun06** [comment]
It is not working as it should. So closing this one and working on a new PR for the issue.

**Pull request closed without merging**